# Project 1

In this first project you will implement a few python functions and generally familiarize yourself with the pandas library.  

Please refer to **numpy-and-pandas.ipynb** noteook in lesson2 and the for loop section of the **python-controlflow.ipynb** notebook in the python_foundations folder. and the pandas documentation [here](http://pandas.pydata.org/pandas-docs/stable/) for assitance. 

I have written the numerical answers you are looking for below - please show me the code you used to generate those answers. 

### Note!  You will need to look within that documentation/ use  other search results on the internet to complete this assignment!

## Question 1: Multiples of Three and Five

If we list all of the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6, and 9. The sum of these multiples is 23. Find the sum of all the multiples of 3 and 5 below 1,000. 

In [10]:
mylist = []
for x in (range(1, 1000)): #range is a function, so you can set start and end by passing different parameters and separating them with a comma.
    if x %3 == 0:
        mylist.append(x)
    elif x %5 == 0:
        mylist.append(x)
print(mylist[:5])

[3, 5, 6, 9, 10]


In [11]:
sum(mylist)

233168

Answer: 233,168


*Note: you may find yourself with the answer 266,333!  Think carefully what is going on with this question and what may be driving the difference between your answer and the correct value! A hint can be found in the control flow notebook.* 

`elif` solves the issue of counting a number that is a multiple of both 3 and 5 twice.

## Question 2: Pandas Intro

In [12]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### 2.1 Load the Citibike-Feb-24 dataset into memory and assign it to the variable "df"

The data are in */data/citibike_feb2014.csv*

Use pd.read_csv function.  Please refer to the documentation [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) if you are having trouble.  

In [13]:
df = pd.read_csv('data/citibike_feb2014.csv')

### 2.2 How many rows and how many columns are there in the dataset?

In [16]:
x = len(df)
y = len(list(df.columns))
z = len(df.columns)

print(x)
print(y)
print(z)

print(df.shape)

224736
15
15
(224736, 15)


A: 224,736 rows, 15 columns

### 2.3 Please print out the first five rows of the dataset

In [17]:
df.head(5)

tripduration            starttime             stoptime  start station id  \
0           382  2014-02-01 00:00:00  2014-02-01 00:06:22               294   
1           372  2014-02-01 00:00:03  2014-02-01 00:06:15               285   
2           591  2014-02-01 00:00:09  2014-02-01 00:10:00               247   
3           583  2014-02-01 00:00:32  2014-02-01 00:10:15               357   
4           223  2014-02-01 00:00:41  2014-02-01 00:04:24               401   

        start station name  start station latitude  start station longitude  \
0      Washington Square E               40.730494               -73.995721   
1       Broadway & E 14 St               40.734546               -73.990741   
2   Perry St & Bleecker St               40.735354               -74.004831   
3       E 11 St & Broadway               40.732618               -73.991580   
4  Allen St & Rivington St               40.720196               -73.989978   

   end station id          end station name  end station latitude  \
0             265  Stanton St & Chrystie St             40.722293   
1             439            E 4 St & 2 Ave             40.726281   
2             251       Mott St & Prince St             40.723180   
3             284     Greenwich Ave & 8 Ave             40.739017   
4             439            E 4 St & 2 Ave             40.726281   

   end station longitude  bikeid    usertype birth year  gender  
0             -73.991475   21101  Subscriber       1991       1  
1             -73.989780   15456  Subscriber       1979       2  
2             -73.994800   16281  Subscriber       1948       2  
3             -74.002638   17400  Subscriber       1981       1  
4             -73.989780   19341  Subscriber       1990       1

### 2.4 What is the average trip duration? (In seconds)

In [18]:
df['tripduration'].mean()

874.5198099102947

In [19]:
df['tripduration'].sum() / len(df)

874.5198099102947

In [20]:
df['starttime_dt'] = df['starttime'].apply(pd.to_datetime)
df['stoptime_dt'] = df['stoptime'].apply(pd.to_datetime)

In [21]:
df['triptime'] = df['stoptime_dt'] - df['starttime_dt']

In [23]:
df['triptime'].head()

0   00:06:22
1   00:06:12
2   00:09:51
3   00:09:43
4   00:03:43
Name: triptime, dtype: timedelta64[ns]

In [24]:
df['triptime_seconds'] = df['triptime'].apply(lambda x: x.total_seconds())

In [25]:
df['triptime_seconds'].mean()

874.5198099102947

A: 874.5198 seconds

### 2.5 What is the total trip duration in this entire dataset in hours?

In [26]:
382 / (6 + 22/60)

60.00000000000001

In [27]:
total_secs = df['tripduration'].sum()
hours = total_secs / 60 / 60
print(hours)

54593.35666666667


In [28]:
hours = df['triptime_seconds'].sum() / 3600
print(hours)

54593.35666666667


In [20]:
df['triptime_hours'] = df['triptime_seconds']/3600
df['triptime_hours'].sum()

54593.356666666674

A: 54593.3567 hours

### 2.6 What is the most popular start station? And how many rides started at that station in Feb 2014?

Note, the pandas cookbook may come in handy for this (look at chapter 1 & 2): https://pandas.pydata.org/pandas-docs/stable/tutorials.html 

In [26]:
list(df.columns)

['tripduration',
 'starttime',
 'stoptime',
 'start station id',
 'start station name',
 'start station latitude',
 'start station longitude',
 'end station id',
 'end station name',
 'end station latitude',
 'end station longitude',
 'bikeid',
 'usertype',
 'birth year',
 'gender',
 'newstarttime',
 'newstoptime',
 'triptime',
 'triptime_seconds',
 'triptime_hours']

In [100]:
df['start station name'].value_counts()

Lafayette St & E 8 St              2920
Pershing Square N                  2719
E 17 St & Broadway                 2493
W 21 St & 6 Ave                    2403
8 Ave & W 31 St                    2171
8 Ave & W 33 St                    1944
W 41 St & 8 Ave                    1916
University Pl & E 14 St            1808
Cleveland Pl & Spring St           1796
Broadway & E 14 St                 1795
Broadway & E 22 St                 1694
E 43 St & Vanderbilt Ave           1688
W 33 St & 7 Ave                    1570
Broadway & W 24 St                 1562
W 31 St & 7 Ave                    1538
Christopher St & Greenwich St      1500
Greenwich Ave & 8 Ave              1500
Great Jones St                     1471
W 13 St & 6 Ave                    1392
W 27 St & 7 Ave                    1391
2 Ave & E 31 St                    1371
Lexington Ave & E 24 St            1363
W 17 St & 8 Ave                    1361
W 18 St & 6 Ave                    1359
E 32 St & Park Ave                 1352


In [103]:
df['start station id'].value_counts()

293     2920
519     2719
497     2493
435     2403
521     2171
490     1944
477     1916
382     1808
151     1796
285     1795
402     1694
318     1688
492     1570
444     1562
379     1538
284     1500
358     1500
229     1471
345     1392
442     1391
528     1371
537     1363
116     1361
168     1359
472     1352
523     1337
426     1324
504     1319
323     1298
505     1290
        ... 
282      136
437      134
343      134
397      130
353      128
389      126
421      121
232      110
245      100
289       94
419       93
120       92
436       91
243       89
144       89
298       89
339       82
242       76
218       76
2001      75
216       72
253       69
372       57
119       57
418       56
431       54
278       45
443       41
2005      36
320        4
Name: start station id, Length: 329, dtype: int64

In [102]:
df['start station id'].value_counts().loc[293]

2920

In [34]:
import datetime

In [39]:
d_ex = datetime.datetime.strptime('2014-02-01 00:00:00', "%Y-%m-%d %H:%M:%S") 

In [40]:
d_ex

datetime.datetime(2014, 2, 1, 0, 0)

In [42]:
d_ex.month

2

In [43]:
df['starttime_dt'][0].month

2

In [45]:
df['month_start'] = df['starttime_dt'].apply(lambda x: x.month)

In [46]:
df['month_start'].value_counts()

2    224736
Name: month_start, dtype: int64

all the months are feb

In [47]:
df2 = df[['start station id','start station name']]

In [56]:
time1 = time.time()
dict2 = {}
for i in range(len(df2)): #to NOT iterate through the df itself, i.e. to NOT iterate through the columns.
    row = df2.iloc[i] #grabs each row of the df
    k = row['start station id']
    if k in dict2.keys(): #this turns the keys of a dict into a list, same can be done for values
        dict2[k] += 1 #if key exists adds by 1
    else:
        dict2[k] = 1 #if key is new, adds key and sets value to 1 (no key can ever have no value)
time2 = time.time()
total_time = time2 - time1
print ("time in seconds: %s" % str(total_time))

time in seconds: 35.311187982559204


In [97]:
time3 = time.time()
dict3 = {}
for u_id in df2['start station id'].unique():
    dict3[u_id] = len(df2[df2['start station id']==u_id])
time4 = time.time()
tot_time = time4 - time3
print('Time in seconds: %s' % str(tot_time))

Time in seconds: 0.304030179977417


In [55]:
time5 = time.time()
print(df2['start station id'].value_counts().index[0])
time6 = time.time()
tot_time = time6 - time5
print('Time in seconds: %s' % str(tot_time))

293
Time in seconds: 0.0047607421875


In [52]:
import time

In [95]:
time.time()

1524245702.8950233

In [59]:
time5 = time.time()
dict4 = df2['start station id'].value_counts().to_dict()
time6 = time.time()
tot_time = time6 - time5
print('Time in seconds: %s' % str(tot_time))

Time in seconds: 0.00567173957824707


In [60]:
import pandas as pd

In [61]:
pd.Series(dict2)

72       676
79       632
82       296
83       301
116     1361
119       57
120       92
127     1019
128     1163
137      508
143      184
144       89
146      530
147      935
150      549
151     1796
152      388
153     1051
157      316
160      633
161     1201
164      650
167      958
168     1359
173      963
174      795
195      713
212     1010
216       72
217      151
        ... 
527     1020
528     1371
529     1007
531      676
532      410
533      354
534      388
536      961
537     1363
538      569
539      352
540     1288
545     1037
546      960
2000     339
2001      75
2002     222
2003    1093
2004     580
2005      36
2006     429
2008     407
2009     439
2010     560
2012     706
2017     359
2021     720
2022     427
2023     330
3002     781
Length: 329, dtype: int64

In [62]:
my_ser = pd.Series(dict2)

In [63]:
my_ser.sort_values(ascending = False, inplace = True)

In [65]:
my_ser.iloc[0]

2920

In [87]:
my_ser.index

Int64Index([ 293,  519,  497,  435,  521,  490,  477,  382,  151,  285,
            ...
             216,  253,  119,  372,  418,  431,  278,  443, 2005,  320],
           dtype='int64', length=329)

In [88]:
my_ser.index[0]

293

In [69]:
row = df2.iloc[0]

In [70]:
row

start station id                      294
start station name    Washington Square E
Name: 0, dtype: object

In [68]:
row['start station id']

294

In [59]:
df['start station id'].to_dict() # a series always has the function `to_dict()`

{0: 294,
 1: 285,
 2: 247,
 3: 357,
 4: 401,
 5: 152,
 6: 325,
 7: 354,
 8: 375,
 9: 285,
 10: 518,
 11: 501,
 12: 388,
 13: 518,
 14: 257,
 15: 477,
 16: 317,
 17: 527,
 18: 504,
 19: 316,
 20: 490,
 21: 518,
 22: 450,
 23: 300,
 24: 474,
 25: 490,
 26: 540,
 27: 347,
 28: 499,
 29: 285,
 30: 403,
 31: 237,
 32: 146,
 33: 497,
 34: 470,
 35: 355,
 36: 540,
 37: 494,
 38: 496,
 39: 237,
 40: 143,
 41: 345,
 42: 368,
 43: 444,
 44: 237,
 45: 497,
 46: 493,
 47: 435,
 48: 229,
 49: 435,
 50: 237,
 51: 526,
 52: 372,
 53: 477,
 54: 349,
 55: 488,
 56: 2012,
 57: 476,
 58: 3002,
 59: 2021,
 60: 443,
 61: 146,
 62: 237,
 63: 351,
 64: 405,
 65: 280,
 66: 490,
 67: 473,
 68: 497,
 69: 345,
 70: 319,
 71: 312,
 72: 404,
 73: 384,
 74: 483,
 75: 527,
 76: 476,
 77: 446,
 78: 453,
 79: 476,
 80: 319,
 81: 116,
 82: 480,
 83: 157,
 84: 432,
 85: 410,
 86: 312,
 87: 503,
 88: 405,
 89: 161,
 90: 326,
 91: 284,
 92: 336,
 93: 497,
 94: 325,
 95: 482,
 96: 432,
 97: 441,
 98: 293,
 99: 394,
 100: 2

A: Station id: 293, Number of rides; 2920

### 2.7 What percentage of the total riders are of usertype "Subscriber"?

In [77]:
df['usertype'].value_counts()

Subscriber    218019
Customer        6717
Name: usertype, dtype: int64

In [78]:
sum(df['usertype'] == 'Subscriber')

218019

In [79]:
subs = sum(df['usertype'] == 'Subscriber') / len(df)
print(subs)

0.9701115976078599


In [85]:
subtype_df = df['usertype'].value_counts().rename('count').to_frame() #this takes a value_counts series, renames it and turns it into a dataframe

In [86]:
subtype_df

count
Subscriber  218019
Customer      6717

In [87]:
subtype_df['ratio'] = subtype_df / len(df2)

In [88]:
subtype_df['ratio'] = subtype_df / subtype_df['count'].sum()

In [89]:
subtype_df

count     ratio
Subscriber  218019  0.970112
Customer      6717  0.029888

A: 97.0112

### What is the average age (in 2014) of the riders in this dataset?  

Note, this requires creating a new column and then taking the difference between 2014 and the rider's birth year, then taking the average!

In [137]:
2014 - df['birth year'][0]

TypeError: unsupported operand type(s) for -: 'int' and 'str'

In [146]:
df['birth year'].head()

0    1991
1    1979
2    1948
3    1981
4    1990
Name: birth year, dtype: object

In [144]:
df['birth year'].astype(int)

ValueError: invalid literal for int() with base 10: '\\N'

In [126]:
df['birth year'].astype(str).astype(int)

ValueError: invalid literal for int() with base 10: '\\N'

In [100]:
df['birth_year'] = pd.to_numeric(df['birth year'])

ValueError: Unable to parse string "\N" at position 31

In [131]:
df['birth year'].loc[31]

'\\N'

In [142]:
df['birth year'].value_counts()

1985    9305
1984    9139
1983    8779
1981    8208
1986    8109
1988    8048
1982    7968
1979    7771
1980    7661
1987    7472
1978    6876
1989    6843
\N      6717
1977    5957
1974    5848
1970    5736
1990    5697
1976    5579
1969    5455
1972    5028
1973    4974
1971    4923
1975    4884
1967    4650
1968    4476
1964    4249
1963    4229
1966    4005
1965    3644
1962    3641
        ... 
1949     725
1946     454
1947     451
1996     334
1944     311
1997     251
1945     214
1942     182
1941     164
1943     130
1940      84
1938      75
1900      68
1939      43
1922      32
1936      31
1937      24
1934      21
1935      14
1901      11
1933      10
1899       9
1932       8
1907       5
1926       5
1910       4
1917       3
1927       2
1921       1
1913       1
Name: birth year, Length: 78, dtype: int64

In [97]:
df[df['birth year'] == '\\N']

tripduration            starttime             stoptime  \
31               664  2014-02-01 00:08:47  2014-02-01 00:19:51   
55               836  2014-02-01 00:16:10  2014-02-01 00:30:06   
222             1277  2014-02-01 01:17:50  2014-02-01 01:39:07   
266            29906  2014-02-01 01:44:59  2014-02-01 10:03:25   
293             2625  2014-02-01 01:56:32  2014-02-01 02:40:17   
302             2289  2014-02-01 02:01:56  2014-02-01 02:40:05   
306              597  2014-02-01 02:02:59  2014-02-01 02:12:56   
329              990  2014-02-01 02:17:02  2014-02-01 02:33:32   
393              735  2014-02-01 03:01:41  2014-02-01 03:13:56   
413             1467  2014-02-01 03:17:49  2014-02-01 03:42:16   
420             1869  2014-02-01 03:30:54  2014-02-01 04:02:03   
432             1017  2014-02-01 03:42:56  2014-02-01 03:59:53   
526             1466  2014-02-01 06:16:45  2014-02-01 06:41:11   
556             1926  2014-02-01 06:42:52  2014-02-01 07:14:58   
615             1572  2014-02-01 07:24:58  2014-02-01 07:51:10   
629              957  2014-02-01 07:31:31  2014-02-01 07:47:28   
788              373  2014-02-01 08:18:27  2014-02-01 08:24:40   
846              264  2014-02-01 08:32:07  2014-02-01 08:36:31   
869             2154  2014-02-01 08:36:25  2014-02-01 09:12:19   
871             2134  2014-02-01 08:36:35  2014-02-01 09:12:09   
911              994  2014-02-01 08:42:27  2014-02-01 08:59:01   
1027            2204  2014-02-01 09:00:35  2014-02-01 09:37:19   
1039             776  2014-02-01 09:02:25  2014-02-01 09:15:21   
1061             292  2014-02-01 09:04:52  2014-02-01 09:09:44   
1093             653  2014-02-01 09:09:42  2014-02-01 09:20:35   
1096            1588  2014-02-01 09:09:55  2014-02-01 09:36:23   
1098            1586  2014-02-01 09:10:00  2014-02-01 09:36:26   
1100            1963  2014-02-01 09:10:13  2014-02-01 09:42:56   
1104             952  2014-02-01 09:10:46  2014-02-01 09:26:38   
1109            1920  2014-02-01 09:11:04  2014-02-01 09:43:04   
...              ...                  ...                  ...   
222820          1737  2014-02-28 18:27:37  2014-02-28 18:56:34   
222848           469  2014-02-28 18:29:14  2014-02-28 18:37:03   
222890           514  2014-02-28 18:32:42  2014-02-28 18:41:16   
222997          1239  2014-02-28 18:41:10  2014-02-28 19:01:49   
222998          1009  2014-02-28 18:41:10  2014-02-28 18:57:59   
223006           744  2014-02-28 18:41:34  2014-02-28 18:53:58   
223053           657  2014-02-28 18:44:25  2014-02-28 18:55:22   
223265           593  2014-02-28 18:59:37  2014-02-28 19:09:30   
223286          1054  2014-02-28 19:01:51  2014-02-28 19:19:25   
223443          6344  2014-02-28 19:15:35  2014-02-28 21:01:19   
223721           752  2014-02-28 19:40:40  2014-02-28 19:53:12   
223759           660  2014-02-28 19:44:09  2014-02-28 19:55:09   
223800           201  2014-02-28 19:49:23  2014-02-28 19:52:44   
223813          1748  2014-02-28 19:50:34  2014-02-28 20:19:42   
223817           506  2014-02-28 19:51:28  2014-02-28 19:59:54   
223847           725  2014-02-28 19:55:44  2014-02-28 20:07:49   
223867          1128  2014-02-28 19:57:58  2014-02-28 20:16:46   
223946           150  2014-02-28 20:08:23  2014-02-28 20:10:53   
223984          1021  2014-02-28 20:14:11  2014-02-28 20:31:12   
224060           401  2014-02-28 20:28:37  2014-02-28 20:35:18   
224129           643  2014-02-28 20:39:41  2014-02-28 20:50:24   
224131           693  2014-02-28 20:40:18  2014-02-28 20:51:51   
224306           356  2014-02-28 21:20:07  2014-02-28 21:26:03   
224317           630  2014-02-28 21:22:25  2014-02-28 21:32:55   
224318           258  2014-02-28 21:22:32  2014-02-28 21:26:50   
224385           280  2014-02-28 21:42:34  2014-02-28 21:47:14   
224438          1106  2014-02-28 22:00:46  2014-02-28 22:19:12   
224525           864  2014-02-28 22:29:52  2014-02-28 22:44:16   
224536           683  2014-02-28 22:32:55  2014-

In [99]:
x = 2014
df['diff_year'] = x - df['birth_year']
df['diff_year'].mean()

TypeError: unsupported operand type(s) for -: 'int' and 'str'

In [151]:
df.iloc[31]

tripduration                                   664
starttime                      2014-02-01 00:08:47
stoptime                       2014-02-01 00:19:51
start station id                               237
start station name                 E 11 St & 2 Ave
start station latitude                     40.7305
start station longitude                   -73.9867
end station id                                 349
end station name           Rivington St & Ridge St
end station latitude                       40.7185
end station longitude                     -73.9833
bikeid                                       17540
usertype                                  Customer
birth year                                      \N
gender                                           0
newstarttime                   2014-02-01 00:08:47
newstoptime                    2014-02-01 00:19:51
triptime                           0 days 00:11:04
triptime_seconds                               664
triptime_hours                 

In [158]:
df[df['birth year']=='\\N'] #need to add an extra \ bc \n is a new line character so \\n = \n in a non new line character form

tripduration            starttime             stoptime  \
31               664  2014-02-01 00:08:47  2014-02-01 00:19:51   
55               836  2014-02-01 00:16:10  2014-02-01 00:30:06   
222             1277  2014-02-01 01:17:50  2014-02-01 01:39:07   
266            29906  2014-02-01 01:44:59  2014-02-01 10:03:25   
293             2625  2014-02-01 01:56:32  2014-02-01 02:40:17   
302             2289  2014-02-01 02:01:56  2014-02-01 02:40:05   
306              597  2014-02-01 02:02:59  2014-02-01 02:12:56   
329              990  2014-02-01 02:17:02  2014-02-01 02:33:32   
393              735  2014-02-01 03:01:41  2014-02-01 03:13:56   
413             1467  2014-02-01 03:17:49  2014-02-01 03:42:16   
420             1869  2014-02-01 03:30:54  2014-02-01 04:02:03   
432             1017  2014-02-01 03:42:56  2014-02-01 03:59:53   
526             1466  2014-02-01 06:16:45  2014-02-01 06:41:11   
556             1926  2014-02-01 06:42:52  2014-02-01 07:14:58   
615             1572  2014-02-01 07:24:58  2014-02-01 07:51:10   
629              957  2014-02-01 07:31:31  2014-02-01 07:47:28   
788              373  2014-02-01 08:18:27  2014-02-01 08:24:40   
846              264  2014-02-01 08:32:07  2014-02-01 08:36:31   
869             2154  2014-02-01 08:36:25  2014-02-01 09:12:19   
871             2134  2014-02-01 08:36:35  2014-02-01 09:12:09   
911              994  2014-02-01 08:42:27  2014-02-01 08:59:01   
1027            2204  2014-02-01 09:00:35  2014-02-01 09:37:19   
1039             776  2014-02-01 09:02:25  2014-02-01 09:15:21   
1061             292  2014-02-01 09:04:52  2014-02-01 09:09:44   
1093             653  2014-02-01 09:09:42  2014-02-01 09:20:35   
1096            1588  2014-02-01 09:09:55  2014-02-01 09:36:23   
1098            1586  2014-02-01 09:10:00  2014-02-01 09:36:26   
1100            1963  2014-02-01 09:10:13  2014-02-01 09:42:56   
1104             952  2014-02-01 09:10:46  2014-02-01 09:26:38   
1109            1920  2014-02-01 09:11:04  2014-02-01 09:43:04   
...              ...                  ...                  ...   
222820          1737  2014-02-28 18:27:37  2014-02-28 18:56:34   
222848           469  2014-02-28 18:29:14  2014-02-28 18:37:03   
222890           514  2014-02-28 18:32:42  2014-02-28 18:41:16   
222997          1239  2014-02-28 18:41:10  2014-02-28 19:01:49   
222998          1009  2014-02-28 18:41:10  2014-02-28 18:57:59   
223006           744  2014-02-28 18:41:34  2014-02-28 18:53:58   
223053           657  2014-02-28 18:44:25  2014-02-28 18:55:22   
223265           593  2014-02-28 18:59:37  2014-02-28 19:09:30   
223286          1054  2014-02-28 19:01:51  2014-02-28 19:19:25   
223443          6344  2014-02-28 19:15:35  2014-02-28 21:01:19   
223721           752  2014-02-28 19:40:40  2014-02-28 19:53:12   
223759           660  2014-02-28 19:44:09  2014-02-28 19:55:09   
223800           201  2014-02-28 19:49:23  2014-02-28 19:52:44   
223813          1748  2014-02-28 19:50:34  2014-02-28 20:19:42   
223817           506  2014-02-28 19:51:28  2014-02-28 19:59:54   
223847           725  2014-02-28 19:55:44  2014-02-28 20:07:49   
223867          1128  2014-02-28 19:57:58  2014-02-28 20:16:46   
223946           150  2014-02-28 20:08:23  2014-02-28 20:10:53   
223984          1021  2014-02-28 20:14:11  2014-02-28 20:31:12   
224060           401  2014-02-28 20:28:37  2014-02-28 20:35:18   
224129           643  2014-02-28 20:39:41  2014-02-28 20:50:24   
224131           693  2014-02-28 20:40:18  2014-02-28 20:51:51   
224306           356  2014-02-28 21:20:07  2014-02-28 21:26:03   
224317           630  2014-02-28 21:22:25  2014-02-28 21:32:55   
224318           258  2014-02-28 21:22:32  2014-02-28 21:26:50   
224385           280  2014-02-28 21:42:34  2014-02-28 21:47:14   
224438          1106  2014-02-28 22:00:46  2014-02-28 22:19:12   
224525           864  2014-02-28 22:29:52  2014-02-28 22:44:16   
224536           683  2014-02-28 22:32:55  2014-

In [157]:
new_df = df[df['birth year']!='\\N']

In [163]:
#new_df['birth year'] = new_df['birth year'].astype(int)
new_df['birth year'] = new_df['birth year'].apply(lambda x: int(x))

C:\Users\ystrano\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [167]:
df['int_birthyear'] = 2014 - new_df['birth year'] #in numpy you can broadcast a calc and it will apply across every row. this runs quickly bc it uses cython (c code) and multi threading 

In [168]:
df['int_birthyear'].mean()

38.50249290199478

In [103]:
df["birth_year"] = df['birth year'].replace('\\N', np.nan) 

adding a \ helps escape the special character issue this string has. the problem with this is that the 0 is going to deflate the average result - this analysis is correct, and that's why we set this to "not a number" and not 0

In [104]:
df.loc[~df['birth_year'].isnull(), 'birth_year'] = df.loc[~df['birth_year'].isnull(), 'birth_year'].astype(int)

In [105]:
df['age_approx'] = 2014 - df['birth_year']

In [107]:
df['age_approx'].mean()

38.50249290199478